### PR TITLE
[WASI-NN] Integrate the OpenVINO runtime

### PR DESCRIPTION
**Pull request by spencergotowork**
Support WasmEdge build integrates OpenVINO.
1. ubuntu build & ci
2. manylinux2014 build  & ci 
3. macos build & ci (there was a problem running the test, cause related problem is WIP)

Plugin loading issues(#2070 ) on macos occur during testing, but wasm files can be executed. windows is also having problems with plugin loading errors. And the code running on windows causes more problems, I am fixing it and plan to publish it on other PR.
The C++ section of this code has been modified in order to compile.

Reference 
[For Linux platforms](https://github.com/WasmEdge/WasmEdge/pull/1988)

### CODE REVIEW

File: .github/workflows/build-extensions.yml [hunk 1]
```diff
@@ -370,26 +398,50 @@ jobs:
...
-            output_bin: libwasmedgePluginWasiCrypto.dylib
-            test_path: test/plugins/wasi_crypto
-            test_bin: wasiCryptoTests
+            output_path: plugins/wasi_nn
+            output_bin: libwasmedgePluginWasiNN.dylib
+            test_path: test/plugins/wasi_nn
+            test_bin: wasiNNTests
+            cmake_plugin_option: "-DWASMEDGE_PLUGIN_WASI_NN_BACKEND=OpenVINO"
             build_type: Release
+          # - name: WASI-Crypto
```
Comment (hydai) [inline .github/workflows/build-extensions.yml:423]: If you want to comment out something, you must leave the reason in the comments.

File: .github/workflows/build-extensions.yml [hunk 2]
```diff
@@ -399,20 +451,28 @@ jobs:
...
           export CC=clang
           export CXX=clang++
-          mkdir -p build
+          cmake -Bbuild -GNinja -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_LINK_LLVM_STATIC=ON -DWASMEDGE_BUILD_PACKAGE="TGZ" ${{ matrix.cmake_plugin_option }} -DWASMEDGE_BUILD_TESTS=ON .          
           cd build
-          cmake -DCMAKE_BUILD_TYPE=${{ matrix.build_type }} -DOPENSSL_ROOT_DIR=/usr/local/opt/openssl -DWASMEDGE_PLUGIN_WASI_CRYPTO=ON -DWASMEDGE_BUILD_TESTS=ON ..
-          cd ${{ matrix.test_path }}
-          make -j
+          ninja wasmedge
+          ls -l tools/wasmedge/wasmedge
```
Comment (hydai) [inline .github/workflows/build-extensions.yml:466]: Why do you need to list the content of the folder `tools/wasmedge/wasmedge`?

File: CMakeLists.txt [hunk 3]
```diff
@@ -136,17 +136,17 @@ set(CPACK_DEBIAN_PACKAGE_SHLIBDEPS ON)
 set(CPACK_DEBIAN_PACKAGE_CONTROL_STRICT_PERMISSION ON)
 set(CPACK_GENERATOR "${WASMEDGE_BUILD_PACKAGE}")
 set(CPACK_PACKAGE_DESCRIPTION "WasmEdge is a high performance, extensible, and hardware optimized WebAssembly Virtual Machine for cloud, AI, and blockchain applications.")
-
+if(WASMEDGE_BUILD_PLUGINS AND WASMEDGE_BUILD_SHARED_LIB)
```
Comment (hydai) [inline CMakeLists.txt:139]: Why do you change the ordering of the plugins?

File: plugins/wasi_nn/wasinnfunc.cpp [hunk 4]
```diff
@@ -243,7 +243,7 @@ Expect<uint32_t> WasiNNLoad::body(const Runtime::CallingFrame &Frame,
     }
 
     // Store the loaded graph.
-    *GraphId = Env.NNGraph.size() - 1;
+    *GraphId = uint32_t(Env.NNGraph.size() - 1);
```
Comment (hydai) [inline plugins/wasi_nn/wasinnfunc.cpp:246]: Use `static_cast` instead.

File: plugins/wasi_nn/wasinnfunc.cpp [hunk 5]
```diff
@@ -375,7 +375,7 @@ Expect<uint32_t> WasiNNInitExecCtx::body(const Runtime::CallingFrame &Frame,
       return static_cast<uint32_t>(WASINN::ErrNo::Busy);
     }
 
-    *Context = Env.NNContext.size() - 1;
+    *Context = uint32_t(Env.NNContext.size() - 1);
```
Comment (hydai) [inline plugins/wasi_nn/wasinnfunc.cpp:378]: Ditto.

File: test/plugins/wasi_nn/CMakeLists.txt [hunk 6]
```diff
@@ -67,7 +61,7 @@ foreach(BACKEND ${WASMEDGE_PLUGIN_WASI_NN_BACKEND})
 endforeach()
 
 target_link_libraries(wasiNNTests
-  PRIVATE
+  PUBLIC
```
Comment (hydai) [inline test/plugins/wasi_nn/CMakeLists.txt:64]: Why do you change this to `PUBLIC`?

File: utils/wasi-nn/install-openvino.sh [hunk 7]
```diff
@@ -11,9 +11,25 @@ fi
...
-echo "deb [signed-by=/usr/share/keyrings/GPG-PUB-KEY-INTEL-OPENVINO-$OPENVINO_YEAR.gpg] https://apt.repos.intel.com/openvino/$OPENVINO_YEAR all main" | tee /etc/apt/sources.list.d/intel-openvino-$OPENVINO_YEAR.list
-apt update
-apt install -y intel-openvino-runtime-ubuntu20-$OPENVINO_VERSION
+
+if cat /etc/*release | grep ^NAME | grep CentOS; then
+  echo "Installing OpenVINO with version ${OPENVINO_VERSION} on CentOS"
+  yum install yum-utils python3 -y
+  yum-config-manager --add-repo https://yum.repos.intel.com/openvino/$OPENVINO_YEAR/setup/intel-openvino-$OPENVINO_YEAR.repo
+  rpm --import https://yum.repos.intel.com/openvino/$OPENVINO_YEAR/setup/RPM-GPG-KEY-INTEL-OPENVINO-$OPENVINO_YEAR
+  # export PATH=$PATH:/opt/python/cp38-cp38/bin
```
Comment (hydai) [inline utils/wasi-nn/install-openvino.sh:20]: Why does this comment out?

**Review by hydai — CHANGES_REQUESTED** [review]

**Comment by q82419** [comment]
Please fix your sign-off to pass the DCO.

**Comment by hydai** [comment]
@spencergotowork 
That's because you are trying to contain others' commits. Could you please use `rebase` instead of `merge` to avoid this issue?

**Comment by hydai** [comment]
Hi @spencergotowork 
Sorry for the delay response. I didn't notice this PR because it keeps draft status.
Do you still want to update this with rebasing to the latest master?

**Comment by hydai** [comment]
That will be awesome. Please let me know if you need any help.

**Comment by q82419** [comment]
WASI-NN plugin and the CI have changed a lot.
Close this and please create another PR for the changes.

**Pull request closed without merging**